### ISSUE TITLE
[feature/question][tgi][udg][software sprites] Is it possible to use TGI to implement cross target software sprites?

### ISSUE THREAD

**Issue by Fabrizio-Caruso** [cc65/cc65#697]
Is it possible to use TGI to implement cross target software sprites?

If the TGI characters could be defined, one could use TGI commands to display software sprites.

Ideally, having specific software sprite APIs would be even better.

I am well aware that software sprites are much slower than hardware sprites but cross target sprites is a very cool idea.

**Comment by oliverschmidt** [comment]
From my POV soft sprites are pretty similiar to bitblit. Support for the latter in TGI has been discussed in the past several times. However, there was no momentum for implementation(s) so far.

The general approach would likely be to extend http://cc65.github.io/doc/sp65.html to be able to create individual binary data suitable for individual TGI drivers. Then a function would be added to the TGI drivers that takes a pointer to such binary data, a coordinate (and maybe something like a ROP aka Raster OPeration).

**Comment by Fabrizio-Caruso** [comment]
As a quick and dirty solution, could one simply redefined fonts and use "text characters" as software sprites?

If possible and convenient for you, please consider implementing the very same APIs for software sprites as the ones in Z88DK, e.g.,
extern void __LIB__ putsprite(int ortype, int x, int y, void *sprite) __smallc;
Lots of games could be ported to CC65 targets instantly.

I am sort of already doing this with my own tool-kit but I am limiting my library to just tile graphics and not pixel-perfect graphics because I want to cover targets with no graphics at all.

**Comment by mrdudz** [comment]
Am Donnerstag, 5. Juli 2018, 13:44:14 CEST schrieb Fabrizio Caruso:
> Is it possible to use TGI to implement cross target software sprites?
> 
> If the TGI characters could be defined, one could use TGI commands to
> display software sprites.
> 
> Ideally, having specific software sprite APIs would be even better.
> 
> I am well aware that software sprites are much slower than hardware sprites
> but cross target sprites is a very cool idea.

cross-target software sprites would be the slowest possible implementation of 
this idea, making it practically useless. even regular software sprites are 
quite tricky to make work reasonably fast even in very machine specific 
implementations - for a start, not using hires mode is a key thing :)

-- 

http://hitmen.eu                 http://ar.pokefinder.org
http://vice-emu.sourceforge.net  http://magicdisk.untergrund.net

Oh, boy, virtual memory! Now I'm gonna make myself a really *big* RAMdisk!

**Comment by Fabrizio-Caruso** [comment]
@mrdudz, yes and no. You are right for some targets (e.g.,  C64).
I am aware that they would be slow. So one would have to just use a few and maybe move them with 2 or 4 pixel steps.
Bit-map software sprites are usable on many targets. Majestic of the Sprites for C16 64K uses very colorful software sprites. Atari 8-bit games ofter resort to software sprites with decent results.
Most VDP targets in Z88DK use software sprites and they are fast enough. My game uses them and they are fast.

**Comment by mrdudz** [comment]
Am Donnerstag, 5. Juli 2018, 16:22:26 CEST schrieb Fabrizio Caruso:
> @mrdudz, yes and no. You are right for some targets (e.g.,  C64).
> I am aware that they would be slow. So one would have to just use a few and
> maybe move them with 2 or 4 pixel steps. Bit-map software sprites are
> usable on many targets. Majestic of the Sprites for C16 64K uses very
> colorful software sprites. Atari 8-bit games ofter resort to software
> sprites with decent results. Most VDP targets in Z88DK use software sprites
> and they are fast enough. My game uses them and they are fast.

they are only "fast enough" if you are a fan of said systems, IMHO. "majestic 
of the sprites" for example is almost unplayable for me, even when the game 
really tries hard to push the limits. its one of those that is SCREAMING it 
was made by demo coders =)

i also like how the reply goes to 4 different addresses now :)

-- 

http://hitmen.eu                 http://ar.pokefinder.org
http://vice-emu.sourceforge.net  http://magicdisk.untergrund.net

I am lucky I am such a bad ass. When my gf whines about all my c64s, i just 
look her in the eyes and say: "Never ask me to choose between them and you". 
<Hedning/G*P>

**Comment by Fabrizio-Caruso** [comment]
@mrdudz it is a matter of applications, if you want super-smooth graphics, you have to resort to system-specific code and abandon C in many cases.
Some games and applications do not need super smooth graphics.
Having cross-platform APIs is very nice for many use-cases but not all.

Some systems have fast bit-map. The C64 is one of the worst in this respect. Nevetheless some simple and less simple games can be made. The Goonies for the C64 uses only software sprites (I don't know why).

**Comment by oliverschmidt** [comment]
> As a quick and dirty solution, could one simply redefined fonts and use "text characters" as software sprites?

Are you aware that the c64 TGI driver doesn‘t support text output at all?

**Comment by Fabrizio-Caruso** [comment]
@oliverschmidt, in the short term I need "software sprites" for other targets (e.g., ATARI_LYNX).
In the longer term, I would like to have software sprites for all targets with bit-mapped graphics.

**Comment by Fabrizio-Caruso** [comment]
@oliverschmidt We could start with something as simple as monochromatic sprites.
If the characters used by text printing functions for TGI can be re-defined, this would provide an initial solution for at least monochromatic sprites at zero cost. Are TGI character user-definable?
Of course, I am proposing something ugly. Some syntactic sugar would make it look nicer.

**Comment by mrdudz** [comment]
i am strongly opposed to adding this kind of stuff to TGI. abusing the font output will not work the way you want it for "sprites" either (while being really really REALLY slow)

**Comment by Fabrizio-Caruso** [comment]
@mrdudz my game uses already TGI text as sprites and it is really fast (tested with the ATARI Lynx target)
Why would a different font set make it slower?
Forget about the abuse... Can one just select a different font set, just for text? Why do you want to prevent people from doing this?

Of course clean software sprite APIs would be a better solution but I don't think this will be implemented any time soon.

**Comment by mrdudz** [comment]
perhaps it can work for *some* targets. somehow. however it definitely wont in the generic case. that disqualifies it for me. (the atari lynx really isnt a good reference, its way faster than other targets)

**Comment by Fabrizio-Caruso** [comment]
@mrdudz some targets do not required lots of moving objects.

Look a other use-cases and not just shoot-em ups or demos.
Some games or applications just require static or semi-static images or one single pointer moving over a static background. Some games do not required single-pixel movements. 
There are lots of use-cases that would benefit from it.

@mrdudz have you looked at monochromatic sprites in Z88DK? They are fast in most targets.
I don't think Z80 systems are faster than 6502 ones.

**Comment by mrdudz** [comment]
a single pointer is covered by the mouse api, i think :)

however, you are already limiting the usecases quite a bit. almost to the point that i'd have to ask "why does it even need sprites, at all". that basically proves the point - its slow and not really useable.

whatever - being able to link a user defined charset is an entire different thing - and that should be possible. i havent used TGI myself for a while, but it shouldnt be hard to just link whatever font you want to use. somehow :)

**Comment by Fabrizio-Caruso** [comment]
@mrdudz using a different character set should be allowed. What people do with it, is their own choice. I would use it also as a cheap and dirty trick to get monochromatic software sprites.

@mrdudz there are many use-cases that do not involve lots of super smoothly moving objects: non action games, action games with few moving elements, anything with a static or slowly moving screen.
Even totally static pictures would benefic from sotware sprites.

In Z88DK they even have smooth software sprites for the very slow Amstrad CPC bit-map mode (it must be slower than C64 due to the amount of memory involved).
Take a look at Z88DK and compile some of their little games. You get the same game from the very same code in more than a dozen systems and the sprites move smoothly. Ain't that cool?

@mrdudz do you propose sothing else to achieve real software sprites in a target-independent way?
I know user-defined characters in text/tile/chunky mode is way better but it is inherently target specific.

**Comment by mrdudz** [comment]
for anything that isnt a basic textmode application - and games in particular - i do infect propose to actually make a target specific program. and that in assembler.

i dont see the point in some game that runs on a dozen platforms - but noone would want to play it because it just cant keep up with target specific games. even the tetris thing i wrote myself was never ment to be a "game" - but an application for testing the respective APIs for that matter.

(Even totally static pictures would benefic from sotware sprites. ? now thats plain nonsense >_<)

**Comment by oliverschmidt** [comment]
> being able to link a user defined charset is an entire different thing - and that should be possible. 

That's already possible. There are two types of fonts:

- A "system" bitmap font either being already present on the machine or being part of the TGI driver. 
- Loadable vector fonts.

I'm not interested in adding support for loadable bitmap fonts! Since many years we have pretty patchy bitmap font support in the TGI drivers. Starting with - but not limited to - the C64 TGI driver not supporting any bitmap font at all. Today bitmap fonts are supposed to be rotatable, scalable, <...>. My idea is to remove all those features. This is supposed to:

- Make the TGI driver smaller which support those features today.
- Ease adding support for a bitmap font to the TGI drivers lacking it today.

From my POV the "system" bitmap font is supposed to deliver basic text output capabilities out-of-the-box. For everything fancy vector fonts should be used.

However, vector font output isn't exactly fast. And I guess it doesn't work well with very small glyphs. Therefore I presume vector fonts aren't interesting for a PLUTO terminal. Together with my statement about not adding support for loadable bitmap fonts this means that I don't see TGI font output as a basis for PLUTO terminal text output.

**Comment by Fabrizio-Caruso** [comment]
@mrdudz this is a matter of taste. A simple little game can be fun and does not have to have arcade-quality graphics to be fun. Some games may have very little eye-candy or no graphics at all and be fun. Tetris is a real game IMHO.

CC65 can be an opportunity to provide a way to create simple cross-target games. Some creative developers could create nice games but not having wants to play such games that only use a common sub-set of the graphics capabilities of many systems.

**Comment by Fabrizio-Caruso** [comment]
@mrdudz, it would be cool to have cross-target APIs CC65 to redefine characters? 
(Z88DK has cross-target APIs for that).

For the ATARI_LYNX target I am currently using TGI (because of no conio support) but I have not found a way to redefine the characters.

**Comment by Fabrizio-Caruso** [comment]
@mrdudz for example vg 5000 has not bitmap mode but you can redefine characters.
Z88DK has huge libs with cross target APIs. I am not asking to make the same design choices. I am just asking for some sort of software sprites. CC65 is a leaner project. Even sound can be made cross-target throught bit banging.

**Comment by mrdudz** [comment]
you are evading the question :)

and no cross-platform "bit banging" sound, please. let the spectrum people damage their ears, i choose different platform for a reason =)

**Comment by Fabrizio-Caruso** [comment]
@mrdudz Vg 5000 is one answer to your question. Z88dk does not use bitmap to draw on it but does provide the same api to redefine its characters. They use api as abstractions for very different things.

Gotta go to sleep!

**Comment by mrdudz** [comment]
> Vg 5000 is one answer to your question.

one exception only proves the rule :)

**Comment by polluks** [comment]
The question is: **Should cc65 become a portable gaming platform?** :)

"The Goonies for the C64 uses only software sprites (I don't know why)." Simple: It's an Apple port.

**Comment by mrdudz** [comment]
yeah same for choplifter, karateka, and a few others. and every single one shows why its not a good idea =)

**Comment by Fabrizio-Caruso** [comment]
@polluks, CC65 is already a portable tool-kit primarily through C and some of its cross-target APIs.
I would like to see CC65 go all the way and become a full cross-platform devkit but this would be A LOT of work and I am not asking for that. @mrdudz would not be happy with it.
Here I was just asking for user-defined characters in TGI and/or some cross-target API to display a given monochromatic image on a bitmap screen (which may be used as a softare sprite or something else).

**Comment by oliverschmidt** [comment]
> Should cc65 become a portable gaming platform?

I don't think that this rhetorical question brings us anywhere in this discussion.

**Comment by oliverschmidt** [comment]
> @oliverschmidt : THIS ticket isnt about PLATO =D

Thanks for pointing out - sorry for mixing up things!

**Comment by oliverschmidt** [comment]
> Even sound can be made cross-target throught bit banging.

I don't see any sound support added to cc65.

**Comment by Fabrizio-Caruso** [comment]
@oliverschmidt, CC65 provides some very limited sound support for the ATMOS target (I use it in my game and library). I am not asking for this. This is OT. I only replied to @mrdudz who wrote it cannot be done.

**Comment by polluks** [comment]
At least I ported a C64 game written in C to a C128 game as a proof of concept
https://csdb.dk/release/?id=162762

**Comment by oliverschmidt** [comment]
> [...] cross target sprites [...]

I'm afraid that this discussion suffers from the same problems every discussion on this topic has suffered from in the past. At least to me it's uncertain that we even talk about the same thing.

I suggested that http://cc65.github.io/doc/sp65.html is used to create target (or rather TGI driver) specific sprites/bitmaps. How does the term "cross target sprites" relate to that?

- You ignore my suggestion.
- You think my suggestion is so worthless that it doesn't even make sense to react to it.
- You associate a different meaning to the term "cross target sprites" then I do.

I don't know but maybe the following helps us to clarify misunderstandings:

I see bitblit (see https://en.wikipedia.org/wiki/Bit_blit) as the base functionality. It can be used in many ways. This isn't related to games. It isn't even related to animation. It could be e.g. used together with vector fonts to create some fancy splash screen with a logo (-> bitblit) and text (-> vector font).

Based on bitblit one can create what I believe is refered to in this discussion as "software sprite". My understanding of this is to draw a non-rectangular pattern and remove it again at a later point.

I presume one can (greatly) improve the performance of software sprites with code that isn't a generic bitblit but optimized for the software sprite. E.g. GEOS for the Apple II has generic bitblit support plus specific software sprite support for drag&drop (plus even more specific software mouse cursor code).

I don't see cc65 as a tool targeting smooth / high performance animations. So from my POV it would be appropriate to have generic bitblit support but no specifc software sprite optimizations.

If the generic bitblit support happens to be fast enough for a specific animation then that's great but that's not a goal per se.

**Comment by Fabrizio-Caruso** [comment]
@oliverschmidt I like your proposal: cross-target bitblit is the way to go IMHO. For high performance, people have to resort to target-specific code.
On the somewhat related question of having TGI text chose the font set (though either API or the linker), I think it would be a nice to have feature considering (I guess) that can be implemented with little effort.
Whatever people choose to do with TGI and bitblit is a matter of their personal taste and style.

**Comment by xlar54** [comment]
if it helps, my project implements a software sprite for the mouse pointer on the c128 VDC:  https://github.com/xlar54/rhml
